specs/logging_implementation.adoc

```diff
@@ -0,0 +1,273 @@
+= Generic message mechanism for logs and alerts in Tendrl"
+
+Logs and alerts need to be treated as messages emitted by Tendrl.
+
+== Problem description
+* The logs, alerts and in flight operations updates in Tendrl are handled differently
+  and there is no standard mechanism which makes sure no internal details are
+  exposed as part of alerts/notifications sent to end user.
+* Ideally logs messages, in flight operation updates and alerts are messages which
+  are emitted from an application.
+
+== Use Cases
+
+Logs and alerts need to be treated as messages.
+
+== Proposed change
+* Logs and alerts need to be treated as messages emitted by Tendrl modules.
+* A schema needs to be defined for emitted messages.
+* Message priorities, types etc. need to be defined at the time of emission.
+* The integration/bridge applications must pass the message as per schema.
+* For messages that serve as notifications to the administrator (or alerts),
+  exact error message must be hardcoded by the developers. No backtraces or
+  system generated error messages should be sent as notifications to the
+  administrator.
+* Notifications sent to the administrator should provide sufficient context to
+  enable debugging of the problem via investigation into the log files.
+* Notifications and alerts need to be delivered to the node agent alert sockets.
+* Common module needs to implement a logging library that hides all the implementation
+  details and configuration options from any of the components using the library.
+  Components should be able to blindly emit messages to the Message Handler module
+  in Tendrl and it should be able to deliver them to either syslog, log files,
+  alert sockets etc. depending upon the metadata and configuration.
+* Syslog must be the default logging configuration. Logging to individual
+  configuration files should be thoroughly discouraged.
+
+== Alternatives
+
+* None
+
+== Data model impact
+
+* Store the operation details in etcd:
+```
+1. messages with inorder_id
+    {job_id}/{in_order_id}
+```
+
+== Impacted modules
+
```

Comment:
We'll have the following priorities:

* Debug: Low level information and implementation details that are never pushed to etcd. This includes backtraces.
* Info: Generic messages. Any messages about flow invocations should be info. Messages originating from storage system or monitoring events may end up here.
* Notice: Hardcoded messages that need to be pushed as notifications. Messages originating from storage system or monitoring events may end up here.
* Warn: Hardcoded messages that are pushed as UI notifications, but contain warnings instead of just information. Messages originating from storage system or monitoring events may end up here.
* Error: Errors generated during execution. Probably as a result of exception handling. These may original from within or outside Tendrl.
* Critical: Operational problems, such as process not responding or hardware failures, critical threshold breaches etc.